src/bases.jl

```diff
@@ -342,26 +337,68 @@ function get_basis(M::Manifold, p, B::ProjectedOrthonormalBasis{:gram_schmidt,
         end
         push!(Ξ, Ξₙ)
         K += 1
-        K * real_dimension(number_system(B)) == dim && return PrecomputedOrthonormalBasis(Ξ)
+        K * real_dimension(number_system(B)) == dim && return CachedBasis(B, Ξ, ℝ)
         @label skip
     end
     @warn "get_basis with bases $(typeof(B)) only found $(K) orthonormal basis vectors, but manifold dimension is $(dim)."
-    return PrecomputedOrthonormalBasis(Ξ)
+    return CachedBasis(B, Ξ)
 end
 
+"""
+    get_vectors(M::Manifold, p, B::AbstractBasis)
+
+Get the basis vectors of basis `B` of the tangent space at point `p`.
+"""
+function get_vectors(M::Manifold, p, B::AbstractBasis)
+    error("get_vectors not implemented for manifold of type $(typeof(M)) a point of type $(typeof(p)) and basis of type $(typeof(B)).")
+end
+get_vectors(M::Manifold, p, B::CachedBasis{<:AbstractBasis,<:AbstractArray}) = B.data
+get_vectors(M::Manifold, p, B::CachedBasis{<:AbstractBasis,<:DiagonalizingBasisData}) = B.data.vectors
+
+#internal for directly cached basis i.e. those that are just arrays – used in show
+_get_vectors(B::CachedBasis{<:AbstractBasis,<:AbstractArray}) = B.data
+_get_vectors(B::CachedBasis{<:AbstractBasis,<:DiagonalizingBasisData}) = B.data.vectors
+
+@doc raw"""
+    hat(M::Manifold, p, Xⁱ)
+
+Given a basis $e_i$ on the tangent space at a point `p` and tangent
+component vector $X^i$, compute the equivalent vector representation
+$X=X^i e_i$, where Einstein summation notation is used:
+
+````math
+∧ : X^i ↦ X^i e_i
+````
+
+For array manifolds, this converts a vector representation of the tangent
+vector to an array representation. The [`vee`](@ref) map is the `hat` map's
+inverse.
+"""
+hat(M::Manifold, x, v) = get_vector(M, x, v, DefaultBasis())
+
+function hat!(M::Manifold, X, p, Xⁱ)
+    is_decorator_manifold(M) === Val(true) && return hat!(base_manifold(M), X, p, Xⁱ)
+    error("hat! operator not defined for manifold $(typeof(M)), array $(typeof(X)), point $(typeof(p)), and vector $(typeof(Xⁱ))")
```

Comment:
I don't see a reason to not have this default:
```suggestion
    return copyto!(X, get_vector(M, p, Xⁱ, DefaultBasis()))
```

Reply:
this would then default to `get_vector!` as soon as it's implemented?

Reply:
I actually implemented is as a fallback to the mutating one.